### PR TITLE
sam0/spi: Don't re-configure SPI device when not needed

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Currently, [`spi_acquire()`](https://github.com/RIOT-OS/RIOT/blob/307faa460a03841101159469777f69a073a429e7/cpu/sam0_common/periph/spi.c#L114) will always re-configure the SPI bus.
If the configuration did not change, this is entirely unnecessary and makes SPI operations take longer than needed.

Instead, save the current configuration and skip the initialization if it didn't change since the last call.

### Testing procedure

All SPI operations should still continue to work as before.

A simple ping test with an spi-connected radio should already show the improvement:

**before:**
```
2019-10-08 13:56:30,361 #  ping6 fe80::d0ae:c1b:2054:58f%7
2019-10-08 13:56:30,373 # 12 bytes from fe80::d0ae:c1b:2054:58f: icmp_seq=0 ttl=64 rssi=-21 dBm time=5.664 ms
2019-10-08 13:56:31,645 # 12 bytes from fe80::d0ae:c1b:2054:58f: icmp_seq=1 ttl=64 rssi=-21 dBm time=5.661 ms
2019-10-08 13:56:32,917 # 12 bytes from fe80::d0ae:c1b:2054:58f: icmp_seq=2 ttl=64 rssi=-21 dBm time=5.641 ms
```

**with this patch:**

```
2019-10-08 13:55:55,398 #  ping6 fe80::d0ae:c1b:2054:58f%7
2019-10-08 13:55:55,414 # 12 bytes from fe80::d0ae:c1b:2054:58f: icmp_seq=0 ttl=64 rssi=-20 dBm time=5.503 ms
2019-10-08 13:55:56,723 # 12 bytes from fe80::d0ae:c1b:2054:58f: icmp_seq=1 ttl=64 rssi=-20 dBm time=5.494 ms
2019-10-08 13:55:57,951 # 12 bytes from fe80::d0ae:c1b:2054:58f: icmp_seq=2 ttl=64 rssi=-20 dBm time=5.521 ms
```

### Issues/PRs references
none

### CODE REVIEW

**Comment by bergzand** [comment]
Nice find, I assume the disable-enable dance takes a significant amount of time.
I have two questions with this:
1. Is there a *significant* performance pentalty with doing the `BAUD`/`CTRLA` register value calculation and comparing them with the current content? This way you don't have to store the current settings in an array and can just compare with the current register content instead.
2. I'm wondering if having the check before the `poweron()` is going to cause subtle bugs in the future. is there a *significant* performance penalty in moving the config check below the `poweron()` call?

**Comment by benpicco** [comment]
Good idea! Better save some bytes of RAM when possible :)
I did not notice any degradation in performance by always performing the calculations and reading the config value from the register instead of RAM.

File: cpu/sam0_common/periph/spi.c [hunk 1]
```diff
@@ -104,29 +104,40 @@ void spi_init_pins(spi_t bus)
...
+
+    /* get exclusive access to the device */
+    mutex_lock(&locks[bus]);
+
+    /* power on the device */
+    poweron(bus);
+
+    /* configuration did not change */
+    if (dev(bus)->BAUD.reg == baud && dev(bus)->CTRLA.reg == ctrla) {
+        return SPI_OK;
```
Comment (dylad) [inline cpu/sam0_common/periph/spi.c]: mutex_unlock missing if you return early

File: cpu/sam0_common/periph/spi.c [hunk 2]
```diff
@@ -104,29 +104,40 @@ void spi_init_pins(spi_t bus)
...
      * NOTE: we could configure the pads already during spi_init, but for
      * efficiency reason we do that here, so we can do all in one single write
      * to the CTRLA register */
-    dev(bus)->CTRLA.reg = (SERCOM_SPI_CTRLA_MODE(0x3) |     /* 0x3 -> master */
-                           SERCOM_SPI_CTRLA_DOPO(spi_config[bus].mosi_pad) |
-                           SERCOM_SPI_CTRLA_DIPO(spi_config[bus].miso_pad) |
-                           (mode <<  SERCOM_SPI_CTRLA_CPHA_Pos));
+    const uint32_t ctrla = SERCOM_SPI_CTRLA_MODE(0x3)       /* 0x3 -> master */
+                         | SERCOM_SPI_CTRLA_DOPO(spi_config[bus].mosi_pad)
+                         | SERCOM_SPI_CTRLA_DOPO(spi_config[bus].mosi_pad)
```
Comment (dylad) [inline cpu/sam0_common/periph/spi.c]:
```suggestion
                         | SERCOM_SPI_CTRLA_DIPO(spi_config[bus].miso_pad)
```

File: cpu/sam0_common/periph/spi.c (hunk 1 above)
Comment (benpicco) [reply cpu/sam0_common/periph/spi.c]: Mutex will be unlocked in `spi_release()`.

Comment (dylad) [reply cpu/sam0_common/periph/spi.c]: Uh good point.

File: cpu/sam0_common/periph/spi.c (hunk 2 above)
Comment (benpicco) [reply cpu/sam0_common/periph/spi.c]: Good catch - fixed!

**Comment by benpicco** [comment]
May I squash?

**Comment by dylad** [comment]
Please go ahead.

**Pull request merged**